enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -213,158 +163,154 @@ the underlying, associated Pod references CSI volumes using `Shares`
 - the `podInfoOnMount` field of the CSI Driver spec triggers the setting of these values; here is
 [an example from the hostpath plugin](https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/deploy/kubernetes-1.17/hostpath/csi-hostpath-driverinfo.yaml#L12)
-- with all this necessary metadata, the CSI driver then
-- reads of the `Secret` or `ConfigMap` noted in the `ProjectedResource`, storing it on disk on initial
-reference to said `Secret` or `ConfigMap`
-- creation of a tmpfs filesystem for each requesting `ServiceAcount` that contains the `Secret` / `ConfigMap` data
-- creation of k8s watches `ProjectedResources`, and for `Secret` / `ConfigMaps` in the namespaces listed in each discovered
-`ProjectedResources`
-- final returning of a volume that is based off of the `ServiceAccount` specific tmpfs for the data noted in the
-`volumeAttributes`
+
+With all this necessary metadata, the CSI driver then:
+
+- Performs a subject access review (SAR) check to verify the pod's service account has permission to GET the referenced Share resource.
+- If the SAR check succeeds, reads of the `Secret` or `ConfigMap` noted in the `ProjectedResource`, storing it on disk on initial reference to said `Secret` or `ConfigMap`.
+- Creates a tmpfs filesystem for each requesting `ServiceAcount` that contains the `Secret` / `ConfigMap` data.
+- Creates k8s watches `ProjectedResources`, and for `Secret` / `ConfigMaps` in the namespaces listed in each discovered `ProjectedResources`.
+- Returns a volume that is based off of the `ServiceAccount` specific tmpfs for the data noted in the `volumeAttributes`.
 
 After initial set up of things, steady state CSI driver activities include:
-- as updates of the watched `Secrets` and `ConfigMaps` come in, SAR checks will be done for each of the previous requesting
-`ServiceAccounts` being tracked against the updated `Secret` and `ConfigMap`
-- for any `ServiceAccounts` that no longer have access, their tmpfs filesystems are unmounted, effectively removing their access to
-the data (though in prototyping and implementation, we'll need focused testing that this scales and mount propagation
-does not become a gating factor)
-- the SAR checks and if need be removal of access to data can facilitate cluster admin scenarios noted in the `User Stories`
-section like credential rotation.
-- But for `ServiceAccounts` that pass the SAR check, the tmpfs file with the data for the `Secret` / `ConfigMap` is updated
 
+- As updates of the watched `Secrets` and `ConfigMaps` come in, SAR checks will be done for each of the previous requesting `ServiceAccounts` being tracked against the updated `Secret` and `ConfigMap`.
+  SAR checks are also done at regular relist intervals, which will pick up permission changes to associated cluster roles/role bindings.
```

Comment:
relist of the Secrets, ConfigMaps, and Shares all result in reexecuting the SARs

Reply:
Do we need this level of detail here?

Reply:
IMO I think the details are relevant in that implication of how it works affects the behavior perceived by users.

Specifically, if permissions are removed, the Share content is *eventually* removed from the Pods affected by the permission change, but not immediately.  There is a time gap between whenever the permission change occurs and the next relist (default 10 or 15 minutes IIRC).

An alternate approach, but one that does not exist yet, aside from *maybe* copying a bunch of internal SCC related code which is not exposed for public consumption, is to listen to changes to clusterroles/clusterrolebindings etc., then reverse engineer whether a clusterrole/clusterrolebinding combination is what allowed a given Pod SA to access a given share.  If we could do this, there would be no perceived delay between permission removal and content removal.

If anything, I think we keep the content here, and perhaps better emphasize the current potential delay here, and this list some form of the nirvana potential future I quickly noted above in the alternatives section.

And to be fair, my original EP that merged last year lacked any exposure to this IMO.